contracts/pallet-ibc/src/ics20_fee/mod.rs

```diff
@@ -0,0 +1,281 @@
+			&mut ctx,
+			output,
+			port_id,
+			channel_id,
+			counterparty_version,
+			relayer,
+		)
+	}
+
+	fn on_chan_open_confirm(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_open_confirm(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_chan_close_init(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_close_init(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_chan_close_confirm(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_close_confirm(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_recv_packet(
+		&self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		packet: &Packet,
+		relayer: &Signer,
+	) -> Result<Acknowledgement, Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		let packet_data =
+			serde_json::from_slice::<PacketData>(packet.data.as_slice()).map_err(|e| {
+				Ics04Error::implementation_specific(format!("Failed to decode packet data {:?}", e))
+			})?;
+		let percent = ServiceCharge::<T>::get().unwrap_or(T::ServiceCharge::get());
+		let fee = percent * packet_data.token.amount.as_u256().as_u128();
+		// Send full amount to receiver
+		let ack = self.inner.on_recv_packet(&mut ctx, output, packet, relayer)?;
+		// Send fee from the receiver's account to the pallet account
+		let receiver = <T as crate::Config>::AccountIdConversion::try_from(packet_data.receiver)
+			.map_err(|_| Ics04Error::implementation_specific("Failed to parse account".to_string()))
+			.expect(
+				"Account Id conversion should not fail, it has been validated in a previous step",
+			);
+		let pallet_account: T::AccountId = T::PalletId::get().into_account_truncating();
+		let pallet_account = IbcAccount(pallet_account);
+		let prefixed_coin = if is_receiver_chain_source(
+			packet.source_port.clone(),
+			packet.source_channel,
+			&packet_data.token.denom,
+		) {
+			let prefix = TracePrefix::new(packet.source_port.clone(), packet.source_channel);
+			let mut c = packet_data.token;
+			c.denom.remove_trace_prefix(&prefix);
+			c.amount = fee.into();
+			c
+		} else {
+			let prefix =
+				TracePrefix::new(packet.destination_port.clone(), packet.destination_channel);
+			let mut c = packet_data.token;
+			c.denom.add_trace_prefix(prefix);
+			c.amount = fee.into();
+			c
+		};
+
+		ctx.send_coins(&receiver, &(pallet_account.into()), &prefixed_coin)
```

Comment:
So we're sending coins to the pallet account. Where do we reduce c.amount by the fee? Initial glance makes it seem like we're forwarding the full amount to the end user.

Reply:
Yeah I forward the entire amount to the receiver's account by default, that's wha the ics20 callback would do, then I forward the fee from the reciever to the pallet account, that way I don't have to repeat  the channel escrow derivation done in the ics20 callback.

Reply:
Gotcha, can you add a comment explaining that?